File: moderate/double_squares.lua
sq, n = {}, 0
for i = 0, 46340 do sq[i*i] = true end
for line in io.lines(arg[1]) do
  if n == 0 then
    n = line
  else
    local x = tonumber(line)
    local num, l, bot, top = 0, {}, math.floor(math.sqrt(x/2)), math.floor(math.sqrt(x))
    for i = bot, top do
      local t = x - i*i
      if sq[t] and not l[t] then
        l[i*i], num = true, num + 1
      end
    end
    print(num)
  end
end

File: moderate/double_squares.rb
sq = {}
n = 0
(0..46_340).each { |i| sq[i * i] = true }

File.open(ARGV[0]).each_line do |line|
  if n == 0
    n = line.to_i
  else
    x = line.to_i
    num = 0
    l = {}
    bot = Math.sqrt(x / 2).to_i
    top = Math.sqrt(x).to_i
    (bot..top).each do |i|
      t = x - i * i
      l[i * i] = true
      num += 1 if sq[t] && !l[t]
    end
    puts num
  end
end
